File: AlgorithmsOnGraphs-UCSD/week4/negative_cycle.cpp
#include <iostream>
#include <vector>

using std::vector;

int negative_cycle(vector<vector<int> > &adj, vector<vector<int> > &cost) {
    //write your code here
    vector<int> distances(adj.size(), INT64_MAX);
    for(int k = 0; k < adj.size() - 1; k++){
        for(int i = 0; i < adj.size(); i++){
            int j = 0;
            for(auto it = adj[i].begin(); it != adj[i].end(); it++){
                if(distances[*it] > distances[i] + cost[i][j]){
                    distances[*it] = distances[i] + cost[i][j];
                }
                else
                    j++;
            }
        }
    }

    for(int i = 0; i < adj.size(); i++){
            int j = 0;
            for(auto it = adj[i].begin(); it != adj[i].end(); it++){
                if(distances[*it] > distances[i] + cost[i][j]){
                    return 1;
                }
                else
                    j++;
            }
        }
    return 0;
}

int main() {
  int n, m;
  std::cin >> n >> m;
  vector<vector<int> > adj(n, vector<int>());
  vector<vector<int> > cost(n, vector<int>());
  for (int i = 0; i < m; i++) {
    int x, y, w;
    std::cin >> x >> y >> w;
    adj[x - 1].push_back(y - 1);
    cost[x - 1].push_back(w);
  }
  std::cout << negative_cycle(adj, cost);
}

File: AlgorithmsOnGraphs-UCSD/week4/negative_cycle.py
#Uses python3

import sys


def negative_cycle(adj, cost):
    #write your code here
    number_of_vertices = len(adj)
    distances = [sys.maxsize] * number_of_vertices
    distances[0] = 0

    for _ in range(number_of_vertices - 1):
        for u in range(number_of_vertices):
            i = 0
            for v in adj[u]:
                if distances[v] > distances[u] + cost[u][i]:
                    distances[v] = distances[u] + cost[u][i]
                else:
                    i += 1
        
    #traversal 2
    for u in range(number_of_vertices):
        i = 0
        for v in adj[u]:
            if distances[v] > distances[u] + cost[u][i]:
                return 1
            else:
                i += 1
    return 0


if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n, m = data[0:2]
    data = data[2:]
    edges = list(zip(zip(data[0:(3 * m):3], data[1:(3 * m):3]), data[2:(3 * m):3]))
    data = data[3 * m:]
    adj = [[] for _ in range(n)]
    cost = [[] for _ in range(n)]
    for ((a, b), w) in edges:
        adj[a - 1].append(b - 1)
        cost[a - 1].append(w)
    print(negative_cycle(adj, cost))
